lib/ace/ext/papeeria/tex_completer.coffee

```diff
@@ -84,16 +92,73 @@ define( (require, exports, module) ->
     value: word
     meta: 'equation'
   )
+  istype = (token, type) -> 
+    return token.type.lastIndexOf(type) > -1
+  
+
+  #we will move this code into tex_completer in next pull request and well make this beautiful
+  init = (editor, bindKey) -> 
+    HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
+    keyboardHandler = new HashHandler();
+    keyboardHandler.addCommand(
+      name: 'add item in list mode',
+      bindKey: bindKey,
+      exec: (editor) ->  
+        pos = editor.getCursorPosition()
+        curLine = editor.session.getLine(pos.row);
+        intendCount = ContextHelper.getNestingOfList(editor.session, pos.row)
+        # it's temporary fix bug with added \item before \begin{itemize|enumerate}
+        if ContextHelper.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
+          editor.insert("\n" + "    ".repeat(intendCount) + "\\item ")
+          return true
+        else 
+          return false
+        
+    );
+    editor.keyBinding.addKeyboardHandler(keyboardHandler);
+
+  class ReferenceGetter
+    constructor: ->
+      @cachedURL =  ""
+      @cache = []
+    getReference: (url, callback) -> 
+      if (url == @cashedURL)
+        return @cashe
+      else 
+        @cachedURL = url
+        json = $.getJSON(url)
+        thisRef = @
+        json.success((data) ->
+          callback(data, thisRef)
+        )
+
+        return @cache
```

Comment:
You will always get empty array from this method, because you will return before (asynchronous) request is executed and completes. Also, if you use promise interface (you do) please use `done` instead of `success` and chain the calls: `getJSON(url).done()` -- this will make the intent clear.